backend/entityservice/database/util.py

```diff
@@ -25,52 +27,103 @@ def query_db(db, query, args=(), one=False):
         try:
-            conn = psycopg2.connect(database=db, user=user, password=pw, host=host)
-        except psycopg2.OperationalError:
-            logger.warning("fallback connecting to default postgres db", database=db, user=user, host=host)
-            conn = psycopg2.connect(database='postgres', user=user, password=pw, host=host)
-
-    except psycopg2.Error:
-        logger.warning("Can't connect to database")
-        raise ConnectionError("Issue connecting to database")
-
-    logger.debug("Database connection established")
-    return conn
+            connection_pool = ThreadedConnectionPool(0, db_max_connections, database=db, user=user, password=pw, host=host)
+        except psycopg2.Error:
+            logger.warning("Can't connect to database")
+            raise ConnectionError("Issue connecting to database")
+    else:
+        logger.warning("The connection pool has already been initialized.")
 
 
-def init_db(delay=0.5):
+@atexit.register
+def close_db_pool():
     """
-    Initializes the database by creating the required tables.
-
-    @param float delay: Number of seconds to wait before executing the SQL.
+    If a connection pool is available, close all its connections and set it to None.
+    This method is always run at the exit of the application, to ensure that we free as much as possible
+    the database connections.
     """
-    time.sleep(delay)
-    db = connect_db()
-    with current_app.open_resource('init-db-schema.sql', mode='r') as f:
-        logger.warning("Initialising database")
-        db.cursor().execute(f.read())
-    db.commit()
+    global connection_pool
+    logger.info("Close the database connection pool")
+    if connection_pool is not None:
+        connection_pool.closeall()
+        connection_pool = None
 
 
 class DBConn:
+    """
+    To connect to the database, it is preferred to use the pattern:
+    with DBConn() as conn:
+        ...
+    It will ensure that the connection is either taken from the connection pool if available and return it to the pool
+    when done, or it will create a single connection which is closed when done.
+    """
+
+    return_to_pool = True
 
     def __init__(self, conn=None):
-        self.conn = conn if conn is not None else connect_db()
+        if conn is None:
+            if connection_pool is not None:
+                logger.debug("Get a database connection from the pool.")
+                try:
+                    self.conn = connection_pool.getconn()
+                except psycopg2.Error:
+                    logger.warning("Can't connect to database")
+                    raise ConnectionError("Issue connecting to database")
+            else:
+                logger.debug("Get a database connection without a pool.")
+                self.__connect_db()
+                self.return_to_pool = False
+
+        else:
+            # Not sure where the connection is coming from...
```

Comment:
this is an artifact that is not needed anymore. delete the optional `conn` parameter and this else clause.

Reply:
:+1: